### PR TITLE
refactor: Move `ui.Theme.from_brand()` Sass code into an `.scss` file

### PR DESCRIPTION
**Pull request by gadenbuie**
The main goal is to move as much Sass code and logic out of Python and into a single `_brand-yml.scss` layer file, building on #1790.

There's a small amount of code insertion that still needs to happen, these places are marked in the Sass file with special comments, e.g. 

```scss
/*-- scss:defaults --*/

/*-- insert(brand.color.palette:defaults) --*/
/*-- insert(brand.defaults:defaults) --*/
/*-- insert(brand.color:defaults) --*/
/*-- insert(brand.typography:defaults) --*/
```

As part of this, we now push a set of brand-specific Sass variables and primarily use these to set other values, e.g. 

```scss
$brand_typography_base_family: Open Sans !default; // optional, added if brand.typography.base.family is present

// ... later ...
$brand_typography_base_family: null !default; // always included to prevent errors

// ... later ...
$font-family-base: $brand_typography_base_family !default; // Set Bootstrap variables
```

Note that we can't use the `null !default` set up for the Bootstrap core variables like `$primary` or `$secondary` etc. because the `null` value ends up infiltrating into places where it's not valid.

### CODE REVIEW

File: shiny/ui/_theme_brand.py [hunk 1]
```diff
@@ -278,35 +195,30 @@ def __init__(
...
-
-        # "Brand" rules (now in forwards order)
-        self._add_rules_brand_colors(css_vars_brand)
-        self._add_sass_brand_rules()
+
+        self._insert_sass("brand.color.palette:defaults", brand_color_palette_defaults)
+        self._insert_sass("brand.defaults:defaults", brand_bootstrap_defaults)
+        self._insert_sass("brand.color:defaults", brand_color_defaults)
+        self._insert_sass("brand.typography:defaults", brand_typography_defaults)
+        self._insert_sass("brand.color.palette:rules", brand_color_palette_rules)
```
Comment (cpsievert) [inline shiny/ui/_theme_brand.py:221]: Any chance this could be done with `.add_defaults()` and `.add_rules()` instead of `._insert_sass()`?

Comment (gadenbuie) [reply shiny/ui/_theme_brand.py:221]: I also like that you can look at the Sass file and have a sense of the pieces that are going to be added. 

Also string substitution isn't elegant but it's certainly easy to do in both Python and R and both implementations will basically be the same thing. It's likely that using the Sass layer mechanisms will work out to look pretty similar in both languages, but it certainly increases the context window of things you need to understand.

In other words, I'm going for a solution that, as much as possible, fits into one Sass file.

Comment (gadenbuie) [reply shiny/ui/_theme_brand.py]: Anyway, that's what I was thinking but I rewrote it to use standard Sass methods and I don't hate it. 4ddf0f51451f98529b8da4785972e587e24537ee

File: shiny/ui/_theme_brand.py [hunk 2]
```diff
@@ -318,48 +236,75 @@ def _get_theme_name(self, brand: "Brand") -> str:
...
+        ```scss
+        $brand_typography_base_size: 16rem;
+        $brand_typography_base_line-height: 1.25;
+        ```
+        """
+        mapped: dict[str, YamlScalarType] = {}
 
         if not brand.typography:
-            return mapped
+            return {}
```
Comment (gadenbuie) [inline shiny/ui/_theme_brand.py:307]:
```suggestion
            return mapped
```

File: shiny/ui/_theme_brand.py [hunk 3]
```diff
@@ -368,167 +313,37 @@ def _prepare_typography_vars(brand: "Brand") -> dict[str, str]:
...
-
             for prop_key, prop_value in prop.items():
-                if prop_key in typography_map[field]:
-                    typo_sass_vars = typography_map[field][prop_key]
-                    for typo_sass_var in typo_sass_vars:
-                        mapped[typo_sass_var] = prop_value
-                else:
-                    warn_or_raise_unmapped_variable(f"typography.{field}.{prop_key}")
+                field = field.replace("-", "_")
+                prop_key = prop_key.replace("-", "_")
```
Comment (cpsievert) [inline shiny/ui/_theme_brand.py:318]: Not sure if I follow 100%, but just noting that it seems those could create a variable name like `$brand_typography_headings_line_height`, when the Sass wants `$brand_typography_headings_line-height`?

File: shiny/www/py-shiny/brand/_brand-yml.scss [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+//*-- brand.color --*//
+$body-color: $brand_color_foreground !default;
+$body-bg-dark: $brand_color_foreground !default;
+$body-bg: $brand_color_background !default;
+$body-color-dark: $brand_color_background !default;
+$body-secondary-color: $brand_color_secondary !default;
+$body-secondary: $brand_color_secondary !default;
+$body-tertiary-color: $brand_color_tertiary !default;
+$body-tertiary: $brand_color_tertiary !default;
```
Comment (cpsievert) [inline shiny/www/py-shiny/brand/_brand-yml.scss:79]: Maybe let's save this for another PR, but from a quick look, it seems `$body-secondary` and `$body-tertiary` don't exist?

File: shiny/www/py-shiny/brand/_brand-yml.scss [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    "white": if(contrast-ratio($white, white) <= 1.15, $white, mix($white, white, 20%)),
+    "black": if(contrast-ratio($black, black) <= 1.15, $black, mix($black, black, 20%))
+  );
+}
+
+/*-- scss:defaults --*/
+
+// Sass variables from `brand` will be inserted (above) here in this order:
+// * brand.color.palette
+// * brand.defaults
```
Comment (cpsievert) [inline shiny/www/py-shiny/brand/_brand-yml.scss:27]:
```suggestion
// * brand.defaults (Brand-defined Bootstrap defaults)
```

File: shiny/ui/_theme_brand.py [hunk 6]
```diff
@@ -278,35 +193,38 @@ def __init__(
...
+            brand_color_defaults,
+            brand_color_palette_rules,
+        ) = ThemeBrand._prepare_color_vars(brand)
+
+        brand_typography_defaults = ThemeBrand._prepare_typography_vars(brand)
+
+        # Defaults ----
+        # Final order is reverse-insertion:
+        # * brand.color.palette
+        # * brand.defaults (Brand-defined Bootstrap defaults)
```
Comment (cpsievert) [inline shiny/ui/_theme_brand.py:212]: I would have expected the ordering here to be swapped (i.e., Bootstrap defaults first since they're the most specific). What's the rationale for the palette coming first?

Comment (gadenbuie) [reply shiny/ui/_theme_brand.py:212]: The goal is to be able to say something like `$navbar-bg: $brand-blue`, which requires this order:

```scss
$brand-blue: "#2233ff";

$navbar-bg: $brand-blue;
```

File: shiny/ui/_theme_brand.py (hunk 3 above)
Comment (gadenbuie) [reply shiny/ui/_theme_brand.py:318]: Great catch, thanks. You're right, I don't want these.

Comment (gadenbuie) [inline shiny/ui/_theme_brand.py:318]:
```suggestion
```

File: shiny/www/py-shiny/brand/_brand-yml.scss (hunk 4 above)
Comment (gadenbuie) [reply shiny/www/py-shiny/brand/_brand-yml.scss:79]: There's a good chance I copied this from the Quarto implementation. I'll check.

File: shiny/ui/_theme_brand.py (hunk 6 above)
Comment (gadenbuie) [reply shiny/ui/_theme_brand.py:212]: Yup, came from here: https://github.com/quarto-dev/quarto-cli/blob/bf550cf7049f19d486b7410cb746b70367d0c359/src/core/sass/brand.ts#L31-L34

Happy to leave this for a follow-up PR.

File: shiny/www/py-shiny/brand/_brand-yml.scss (hunk 4 above)
Comment (gadenbuie) [reply shiny/www/py-shiny/brand/_brand-yml.scss:79]: Yup, came from here: https://github.com/quarto-dev/quarto-cli/blob/bf550cf7049f19d486b7410cb746b70367d0c359/src/core/sass/brand.ts#L31-L34

Happy to leave this for a follow-up PR.

File: shiny/ui/_theme_brand.py (hunk 6 above)
Comment (cpsievert) [reply shiny/ui/_theme_brand.py:212]: Ah, ok, that feels worth it. A mention of that capability in the Sass would be great.

Comment (gadenbuie) [reply shiny/ui/_theme_brand.py:212]: idk this feels a lot like the kind of comment that's explained by the code. `brand.color.palette` is only the brand's named colors and it's Sass variables come before `brand.defaults` because otherwise they wouldn't be usable. If it were the other way around, I think it'd need to be commented.

File: shiny/ui/_theme_brand.py [hunk 7]
```diff
@@ -318,45 +236,72 @@ def _get_theme_name(self, brand: "Brand") -> str:
...
+            # brand.color.primary = "#007bff"
+            # ==> $brand_color_primary: #007bff !default;
+            # ==> $primary: $brand_color_primary !default;
+
+            brand_color_var = f"brand_color_{thm_name}"
+            defaults_dict[brand_color_var] = thm_color
+            # Currently, brand.color fields are directly named after Bootstrap vars. If
+            # that changes, we'd need to use a map here. These values can't be set to
+            # `null !default` because they're used by maps in the Bootstrap mixins layer
+            # and cause errors if a color is `null` rather than non-existent.
```
Comment (cpsievert) [inline shiny/ui/_theme_brand.py:261]: Could it be that this is the problem? https://github.com/rstudio/bslib/blob/112efd0e/inst/lib/bs5/scss/_variables.scss#L303

Comment (gadenbuie) [reply shiny/ui/_theme_brand.py:261]: Yup! Fixed in rstudio/bslib#1149

**Review by cpsievert — APPROVED** [review]
